### PR TITLE
feat: BREAKING_CHANGE rewrite in typescript

### PR DESCRIPTION
**Pull request by popomore**
This Pull Request is rewriting code in Typescript

### ES Module Import

use default export 

```typescript
import is from 'is-type-of';
is.string('');
```

compatible with commonjs

```javascript
const is = require('is-type-of');
const { string } = require('is-type-of');
is.string('');
string('');
```

use checking method directly, using `isString('')` is better than `string('')`

```typescript
import { isString } from 'is-type-of';
isString('');
```

### Type Guard

Accroding to https://www.typescriptlang.org/docs/handbook/2/narrowing.html, support type checking

```typescript
const val: unknown = '1';
val.includes('1');    // error: val is of type 'unknown' 
if (isString(val)) {
  val.includes('1');  // success
}
```

### New API

- [ ] isNullable
- [ ] isInteger
- [ ] isInteger32
- [ ] isAsyncGeneratorFunction
- [ ] isPromiseLike

## Breaking Change

- drop Node 14
- Don't depend on core-util-is, isstream

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,18 +34,21 @@
...
     "egg-bin": "^6.4.0",
+    "eslint-config-egg": "^12.2.1",
     "long": "^3.2.0",
     "semver": "^5.4.1",
-    "tsd": "^0.28.1"
+    "ts-expect": "^1.3.0",
+    "typescript": "^5.1.6"
+  },
+  "dependencies": {
+    "@types/long": "^5.0.0"
```
Comment (fengmk2) [inline package.json:52]: dependencies 不需要加

File: src/types/primitive.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
+// https://developer.mozilla.org/en-US/docs/Glossary/Primitive
+
+type Primitive = string | number | bigint | boolean | symbol | null | undefined;
+type Nullable = null | undefined;
+
+export function isString(val?: unknown): val is string {
```
Comment (fengmk2) [inline src/types/primitive.ts:6]: `val is string` 学习了

File: src/types/primitive.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    typeof val === 'boolean' ||
+    typeof val === 'number' ||
+    typeof val === 'string' ||
+    typeof val === 'symbol' ||
+    typeof val === 'bigint' ||
+    typeof val === 'undefined';
+}
+
+// Number Extension
+export function isInteger(val?: unknown): val is number {
```
Comment (fengmk2) [inline src/types/primitive.ts:49]: isBigint 是不是也加上？

File: src/types/util.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+export function isInstanceOf<T extends Class>(val: unknown, Clazz: T): val is InstanceType<T> {
+  return val instanceof Clazz;
+}
+
+export function isPrototypeOf<T extends Class>(val: unknown, Clazz: T): val is InstanceType<T> {
+  return val instanceof Clazz;
+}
+
+export function hasOwnProperty<T extends object>(obj: T, prop: PropertyKey): obj is T & Record<typeof prop, unknown> {
```
Comment (fengmk2) [inline src/types/util.ts:12]: 这体操好高级

File: src/types/primitive.ts (hunk 3 above)
Comment (popomore) [reply src/types/primitive.ts:49]: 加上了，bigint 也是基础类型。

File: src/types/primitive.ts (hunk 2 above)
Comment (popomore) [reply src/types/primitive.ts:6]: 断言的时候用 asserts 也很好用，类型确定 https://blog.logrocket.com/assertion-functions-typescript/

**Comment by popomore** [comment]
文档也写完了

**Review by fengmk2 — APPROVED** [review]

**Comment by fengmk2** [comment]
2.0.0 released

**Pull request merged**